source/extensions/filters/common/rbac/config.cc

```diff
@@ -0,0 +1,65 @@
+#include "extensions/filters/common/rbac/config.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace RBAC {
+
+#define INIT_RBAC_ENGINES(config, disable_http_rules)                                              \
+  engine_(config.has_rules() ? absl::make_optional<RoleBasedAccessControlEngineImpl>(              \
+                                   config.rules(), disable_http_rules)                             \
+                             : absl::nullopt),                                                     \
+      shadow_engine_(config.has_shadow_rules()                                                     \
+                         ? absl::make_optional<RoleBasedAccessControlEngineImpl>(                  \
+                               config.shadow_rules(), disable_http_rules)                          \
+                         : absl::nullopt)
+
+RoleBasedAccessControlRouteSpecificFilterConfig::RoleBasedAccessControlRouteSpecificFilterConfig(
+    const envoy::config::filter::http::rbac::v2::RBACPerRoute& per_route_config)
+    : INIT_RBAC_ENGINES(per_route_config.rbac(), false /* disable_http_rules */) {}
+
+RoleBasedAccessControlFilterStats
+RoleBasedAccessControlFilterConfig::generateStats(const std::string& prefix, Stats::Scope& scope) {
+  const std::string final_prefix = prefix + "rbac.";
+  return {ALL_RBAC_FILTER_STATS(POOL_COUNTER_PREFIX(scope, final_prefix))};
+}
+
+RoleBasedAccessControlFilterConfig::RoleBasedAccessControlFilterConfig(
+    const envoy::config::filter::http::rbac::v2::RBAC& proto_config,
+    const std::string& stats_prefix, Stats::Scope& scope)
+    : stats_(generateStats(stats_prefix, scope)),
+      INIT_RBAC_ENGINES(proto_config, false /* disable_http_rules */) {}
```

Comment:
Why do you need the `INIT_RBAC_ENGINES` macro? You can call the other constructors here. Or extract common part to other class, commented in header.

Reply:
The main reason is the type of `proto_config` are different.
For `RoleBasedAccessControlRouteSpecificFilterConfig`, the type is `RBACPerRoute`;
For `RoleBasedAccessControlFilterConfig`, the type is `http::rbac::v2::RBAC` and `network::rbac::v2::RBAC` respectively.
Do you prefer a template function instead of the `INIT_RBAC_ENGINES` macro?

Reply:
yeah I would prefer a template function

Reply:
Done.